slick/src/main/scala/slick/jdbc/RowsPerStatement.scala

```diff
@@ -0,0 +1,11 @@
+package slick.jdbc
+
+sealed trait RowsPerStatement
+
+object RowsPerStatement {
+  /** A single statement with all rows should be used */
+  case object All extends RowsPerStatement
+
+  /** A separate statement is used for each row */
+  case object One extends RowsPerStatement
```

Comment:
I understand that this is generally a good way to provide type safety in Scala.
However, each Profile may have different options that should be supported.
I believe Slick has an architecture that basically provides APIs in JdbcProfile and different ones for different database products are provided in each Profile.
This implementation forces each Profile to have only two types of implementations.

Reply:
I achieved this in https://github.com/slick/slick/commit/7e60b4369e1c6aa2be3ab80c4e4331ded37465de (which I could add to the branch, or PR separately after this is merged) by using a type member.

https://github.com/slick/slick/commit/7e60b4369e1c6aa2be3ab80c4e4331ded37465de#diff-fef81a55b3619e8e8d49ed29a91ec391c8a7d87e76782212f885c30958786822

It's true that even there, I did it to accommodate Oracle specifically, not in a fully open way, but it could be tweaked a little to achieve that.

tl;dr type members + abstract methods allow refining a subset of a sealed trait hierarchy.

That said I don't recall adding `sealed`, so I'm unclear if your critique is of `sealed`, of my having moved the hierarchy to top-level, or something else.

I moved it to top-level to make imports easier and more consistent with other things in Slick.

Reply:
Oh, I thought I had removed sealed, but it remained.